.github/workflows/golangci-lint.yml

```diff
@@ -1,22 +1,19 @@
-name: golangci-lint
-on:
-  push:
-    tags:
-      - v*
-    branches:
-      - master
-      - 'release/**'
-  pull_request:
+name: ci-golang-lint
+on: [push, pull_request]
 jobs:
-  golangci:
+  golang-lint:
     name: lint
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - name: golangci-lint
+      - name: golang-lint
+        env:
+          # prevent OOM
+          GOGC: 10
         uses: golangci/golangci-lint-action@v2
```

Comment:
Maybe we should add a lint target to the main Makefile that runs https://github.com/golangci/golangci-lint?

There doesn't seem to be any linting in place. Would solve a common issue where someone pushes but CI fails due to a simple linter issue.

Reply:
https://golangci-lint.run/usage/install/#local-installation <- I was thinking of adding some steps for this. 

Unfortunately the 21.03 cut has always had `lint` failures (on & off). The easier approach would be to have this job run and allow this to fail temporarily (I know, highly undesirable). And then we can fix it up for local/remote run under a separate `issue`. The challenge here is to get our `CI` & `tests` in order first. This is a `chicken-&-egg` problem in a way, because - attempting to fix `linter` reported issues will need a decent validation via CI (IMO). As it stands, the `linter` fixes are better put off till the `unit/integ` tests are made to work.

Also another thing I am considering is https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks (pre-commit hooks on developer machines). This would run `lint` / `formatters` etc. before push. This can be introduced once we address the last point I mentioned.

Reply:
> Unfortunately the 21.03 cut has always had lint failures (on & off)

I took a stroll through the lint output. Some errors can/should be squashed with //nolint:xxx tags in the source. Others, such as those from `gosec` should be taken pretty darn seriously. If @akon-dey agrees, I could take those and the other less serious issues on this week to get the linters passing without having to disable in CI

Reply:
> I am considering is https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks

I'm not a fan of this as it might seriously hinder contributions from the public. IMO, requiring a standard and commonly used linter be run before building/pushing is the correct method.

Reply:
>seriously hinder contributions from the publicI agree with your second point around pre-commit hooks as they can

Reply:
> it might seriously hinder contributions from the public

Reply:
### `pre-commit` hooks
> might seriously hinder contributions from the public

I agree with you here ^ - `pre-commit` hooks can slow the setup process by introduce an additional step outside of our `make` files.

### `go lint` discussion
- you can look at all the failures here https://github.com/dgraph-io/dgraph/runs/7961191027?check_suite_focus=true#step:3:35 (this is a past run on a branch off of `main`)
- we could skip a few, and address the `gosec` stuff as you indicated. I am a little worried while touching some errors like these https://github.com/dgraph-io/dgraph/runs/7961191027?check_suite_focus=true#step:3:112 <- as this (may) need a CI post fix. (but these are not a lot, and are countable)


https://github.com/dgraph-io/dgraph/runs/7961191027?check_suite_focus=true#step:3:35

Reply:
> I am a little worried

Most of these gosimple's are really stylistic changes -- often times the generated assembler is no different. And provided we have unit tests for the majority, really not concerning to me.

Reply:
awesome ^ +1. Thanks for clarifying that ... that's helpful to know 👍. @akon-dey - @matthewmcneely has a game plan to help with the `linter` issues we are seeing. It would be good to get help here. We can probably discuss this outside of this PR as well.